source/simulator/stokes_matrix_free.cc

```diff
@@ -443,20 +445,34 @@ namespace aspect
 
             if (cell_data->enable_newton_derivatives)
               {
-                const SymmetricTensor<2,dim,VectorizedArray<number>> grads_phi_u_i = velocity.get_symmetric_gradient (q);
+                VectorizedArray<number> deta_deps_times_grads_phi_u(0.);
+                VectorizedArray<number> eps_times_grads_phi_u(0.);
+                VectorizedArray<number> deta_dp_times_phi_p(0.);
+                for (const unsigned int r : velocity.quadrature_point_indices())
+                  {
+                    deta_deps_times_grads_phi_u += cell_data->newton_factor_wrt_strain_rate_table(cell,r)
+                                                   * grads_phi_u[r];
+                    deta_dp_times_phi_p += cell_data->newton_factor_wrt_pressure_table(cell,r)
+                                           * phi_p[r];
+                    if (cell_data->symmetrize_newton_system)
+                      eps_times_grads_phi_u += cell_data->strain_rate_table(cell,r)
+                                               * grads_phi_u[r];
+                  }
 
-                SymmetricTensor<2,dim,VectorizedArray<number>> newton_velocity_term =
-                  (grads_phi_u_i * cell_data->strain_rate_table(cell,q))
-                  * cell_data->newton_factor_wrt_strain_rate_table(cell,q);
+                const SymmetricTensor<2,dim,VectorizedArray<number>> viscosity_derivative_term =
+                  ( cell_data->symmetrize_newton_system ?
+                    ( cell_data->strain_rate_table(cell,q) * deta_deps_times_grads_phi_u +
+                      cell_data->newton_factor_wrt_strain_rate_table(cell,q) * eps_times_grads_phi_u ) :
+                    2. * cell_data->strain_rate_table(cell,q) * deta_deps_times_grads_phi_u )
+                  +
+                  2. * cell_data->strain_rate_table(cell,q) * deta_dp_times_phi_p;
 
-                if (cell_data->symmetrize_newton_system)
-                  newton_velocity_term +=
-                    (cell_data->newton_factor_wrt_strain_rate_table(cell,q)*grads_phi_u_i)
-                    * cell_data->strain_rate_table(cell,q);
-                velocity.submit_symmetric_gradient(sym_grad_u + newton_velocity_term, q);
+                velocity.submit_symmetric_gradient(sym_grad_u + viscosity_derivative_term, q);
               }
             else
               velocity.submit_symmetric_gradient(sym_grad_u, q);
+
+            pressure.submit_value(-cell_data->pressure_scaling * div, q);
```

Comment:
I have followed the conventional sequence velocity -> pressure when rewriting this part... It is better to preserve the original sequence since the terms to be submitted to pressure has already been done far above. However, the naming of variables in the original code is a little bit confusing, so I rewrite it again to make it more readable.

It is worth noting that my modification is simple but not efficient. The function local_apply() is called in each and every vector multiplication of the CG solver, so it is inefficient to take another loop over the quadrature points in this function. A better way is to store the averages of newton derivatives in cell_data, which is only calculated once in each nonlinear iteration. But I do not know how to implement this idea yet.